Themes and Posts State: use withPersistence to serialize QueryManagers

jsnajdr opened the pull request:
This PR updates the Themes and Posts reducers, which both use a `*QueryManager` to represent the state, to serialize them using the `withPersistence` helper.

It also removes several instances of `withoutPersistence` from the Themes reducer.

Tip for reviewers: when ignoring whitespaces (`git diff -w`) this diff is much shorter and easier to read.

Spinoff from #50222.

**How to test:**
All the affected functionality is covered by unit tests.

jsnajdr committed:
Posts State: use withPersistence to de/serialize PostQueryManager

jsnajdr committed:
Themes State: use withPersistence to de/serialize ThemeQueryManager

tyxla reviewed (commented):
Some persistence tests seem to be failing here.

jsnajdr committed:
Fix how serialized initial state is calculated

jsnajdr commented:
> Some persistence tests seem to be failing here.

Fixed in 56a693c.

The `withSchemaValidation` helper calculates reducer's serialized initial state by running:
```js
const serializedInitialState = reducer( undefined, { type: SERIALIZE } );
```

But the `withPersistence` helper doesn't support passing `undefined` through the serialization handler. Consider these "before" and "after" reducers:

Before:
```js
const reducer = ( state = [], action ) => {
  switch ( action.type ) {
    case PUSH: return [ ...state, action.item ];
    case SERIALIZE: return state.filter( Boolean );
    default: return state;
  }
};
```

After:
```js
const innerReducer = ( state = [], action ) => {
  switch ( action.type ) {
    case PUSH: return [ ...state, action.item ];
    default: return state;
  }
};

const reducer = ( state, action ) => {
  switch ( action.type ) {
    case SERIALIZE: return state.filter( Boolean );
    default: return innerReducer( state, action );
  }
}
```

The second example is how `withPersistence` effectively adds the `SERIALIZE` handler. But the difference is that in the second example, the `state` doesn't get a default value before being passed to the `SERIALIZE` handler. That's why the `state.filter()` call will crash.

And that's what's been happening in our tests.

Fixed by improving the `serializedInitialState` calculation and adding one extra test for `withSchemaValidation` to catch the bug.

tyxla reviewed (approved):
Looks good to me 👍
